tools/autograd/templates/Functions.cpp

```diff
@@ -1238,88 +1243,176 @@ static inline int64_t _min_storage_size(IntList sizes, IntList strides, int64_t
       out_strides_.insert(out_strides_.begin(), stride_i);
     }
   }
-  // Step (2)~(4) for the algorithm in NOTE [ Detecting Memory Overlap Within A Strided Tensor ]
-  //              on output geometry
-  auto out_maybe_overlap = _maybe_overlapping_memory(out_sizes_, out_strides_);
+  if (!out_expanded_dims.empty()) {
+    grad = grad.sum(out_expanded_dims, false);
+  }
+  grad = grad.squeeze();
 
   // For input geometry,
-  //   check for size 0 dimensions,
-  //   skip size 1 dimensions,
-  // Step (0)~(1) for the algorithm in NOTE [ Detecting Memory Overlap Within A Strided Tensor ]
-  //              on input geometry
-  auto idim = input_geometry.dim();
+  //   check for size 0 dimensions,                     (Alg_overlap_inp step 0)
+  //   skip size 1 dimensions,                          (Alg_overlap_inp step 1)
+  //   div grad by \prod size_of_expanded_dims          (Alg_bwd step 0.b)
+  auto inp_dim = input_geometry.dim();
   IntList inp_sizes = input_geometry.sizes(), inp_strides = input_geometry.strides();
-  std::vector<int64_t> inp_sizes_, inp_strides_;
-  inp_sizes_.reserve(idim);
-  inp_strides_.reserve(idim);
-  for (int64_t i = idim - 1; i >= 0; i--) {
+  int64_t inp_multiplicity = 1;
+  std::vector<int64_t> inp_sizes_, inp_strides_, inp_sizes_w_expanded_dims;
+  inp_sizes_.reserve(inp_dim);
+  inp_strides_.reserve(inp_dim);
+  inp_sizes_w_expanded_dims.reserve(out_dim);
+  for (int64_t i = inp_dim - 1; i >= 0; i--) {
     auto size_i = inp_sizes[i];
     auto stride_i = inp_strides[i];
     if (size_i == 0) {
-      return at::zeros(input_geometry.sizes(), grad.type());
-    } else if (size_i != 1) {
-      inp_sizes_.insert(inp_sizes_.begin(), size_i);
-      inp_strides_.insert(inp_strides_.begin(), stride_i);
+      return at::zeros(input_geometry.sizes(), ty);
+    } else if (stride_i == 0) {
+      // Calculate multiplicity given by expanded dims.  (Alg_bwd step 0.b)
+      inp_multiplicity *= size_i;
+      // Moreover, we maintain a size where expanded dims are changed to size 1
+      // dims so we can easily .view(...).expand(...) on the returned grad, if
+      // neede, for Alg_bwd step 0.c.
+      inp_sizes_w_expanded_dims.insert(inp_sizes_w_expanded_dims.begin(), 1);
+    } else  {
+      if (size_i != 1) {
+        inp_sizes_.insert(inp_sizes_.begin(), size_i);
+        inp_strides_.insert(inp_strides_.begin(), stride_i);
+      }
+      inp_sizes_w_expanded_dims.insert(inp_sizes_w_expanded_dims.begin(), size_i);
     }
   }
-  // Step (1)~(4) for the algorithm in NOTE [ Detecting Memory Overlap Within A Strided Tensor ]
-  //              on input geometry
-  auto inp_maybe_overlap = _maybe_overlapping_memory(inp_sizes_, inp_strides_);
+  // Divide grad by multiplicity given by expanded dims.  (Alg_bwd step 0.b)
+  if (inp_multiplicity > 1) {
+    grad = grad.div(inp_multiplicity);
+  }
+
+  // Check output strides & sizes.  (Alg_overlap_out steps 2 ~ 4)
```

Comment:
Is there any reason you ended up inlining `_maybe_overlapping_memory` here?

Reply:
Yes :) Because I end up doing extra things in the original loop in `_maybe_overlapping_memory`. And these extra things are different for input & output.